### PR TITLE
feat(server): Machine learning's image optimisations

### PR DESCRIPTION
**Pull request by ollywelch**
I've made a few improvements to the ML container image. They are
- Using gunicorn to run the server, not Flask
- Building the venv and installing dependencies in the python:3.10 image, but using the python:3.10-slim image to run the server. This results in a ~200MB image. (https://hub.docker.com/layers/ollywelch/immich_machine_learning/latest/images/sha256-6453b78db4d1cfb8eea3993fa364c3b27a03b3a519b18f3913460f5f7004f96d?context=explore)
- Install the python requirements from a requirements.txt file.

### CODE REVIEW

File: machine-learning/gunicorn.conf.py [hunk 1]
```diff
@@ -0,0 +1,28 @@
+"""
+Gunicorn configuration options.
+https://docs.gunicorn.org/en/stable/settings.html
+"""
+import os
+
+
+# Set the bind address based on the env
+port = os.getenv('MACHINE_LEARNING_PORT') or "3003"
+bind = f"0.0.0.0:{port}"
```
Comment (bo0tzz) [inline machine-learning/gunicorn.conf.py:10]: Is it possible to configure this so it will also bind on ipv6 if that is available?

**Review by bo0tzz — APPROVED** [review]
Good stuff, thank you!

File: machine-learning/gunicorn.conf.py (hunk 1 above)
Comment (ollywelch) [reply machine-learning/gunicorn.conf.py:10]: Ill have a look

File: machine-learning/Dockerfile [hunk 2]
```diff
@@ -1,18 +1,26 @@
...
+    PYTHONDONTWRITEBYTECODE=1 \
+    PYTHONUNBUFFERED=1 \
+    PATH="/opt/venv/bin:$PATH"
+
+WORKDIR /usr/src/app
 
 COPY . .
 
-CMD ["python", "src/main.py"]
+CMD ["gunicorn", "src.main:server"]
```
Comment (jrasm91) [inline machine-learning/Dockerfile:26]: Can you update the dev docker compose to use python still? Otherwise we don't have live reload during development.

Comment (ollywelch) [reply machine-learning/Dockerfile:26]: Looks like it is already there in the compose file:
https://github.com/immich-app/immich/blob/main/docker/docker-compose.dev.yml#L33

Comment (jrasm91) [reply machine-learning/Dockerfile:26]: Oh perfect! I didn't think we used that there.

**Review by jrasm91 — APPROVED** [review]
Looks good! Did you validate the image works as expected and runs correctly with this way of installing dependencies?

File: machine-learning/gunicorn.conf.py (hunk 1 above)
Comment (ollywelch) [reply machine-learning/gunicorn.conf.py]: Thought it was best to just make this configurable in the environment - people can use IPv6 if they need to that way

**Comment by ollywelch** [comment]
@jrasm91 I've built it on my rpi and ran a detect objects job and it seems to be working well!

**Comment by jrasm91** [comment]
Perfect! Thanks

**Comment by alextran1502** [comment]
After merging and having it build on main so I can test, I encountered errors while starting the container

```
immich_machine_learning    | Traceback (most recent call last):
immich_machine_learning    |   File "/opt/venv/lib/python3.10/site-packages/torch/__init__.py", line 172, in _load_global_deps
immich_machine_learning    |     ctypes.CDLL(lib_path, mode=ctypes.RTLD_GLOBAL)
immich_machine_learning    |   File "/usr/local/lib/python3.10/ctypes/__init__.py", line 374, in __init__
immich_machine_learning    |     self._handle = _dlopen(self._name, mode)
immich_machine_learning    | OSError: libcublas.so.11: cannot open shared object file: No such file or directory
immich_machine_learning    |
immich_machine_learning    | During handling of the above exception, another exception occurred:
immich_machine_learning    |
immich_machine_learning    | Traceback (most recent call last):
immich_machine_learning    |   File "/opt/venv/lib/python3.10/site-packages/transformers/utils/import_utils.py", line 1110, in _get_module
immich_machine_learning    |     return importlib.import_module("." + module_name, self.__name__)
immich_machine_learning    |   File "/usr/local/lib/python3.10/importlib/__init__.py", line 126, in import_module
immich_machine_learning    |     return _bootstrap._gcd_import(name[level:], package, level)
immich_machine_learning    |   File "<frozen importlib._bootstrap>", line 1050, in _gcd_import
immich_machine_learning    |   File "<frozen importlib._bootstrap>", line 1027, in _find_and_load
immich_machine_learning    |   File "<frozen importlib._bootstrap>", line 1006, in _find_and_load_unlocked
immich_machine_learning    |   File "<frozen importlib._bootstrap>", line 688, in _load_unlocked
immich_machine_learning    |   File "<frozen importlib._bootstrap_external>", line 883, in exec_module
immich_machine_learning    |   File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
immich_machine_learning    |   File "/opt/venv/lib/python3.10/site-packages/transformers/pipelines/__init__.py", line 49, in <module>
immich_machine_learning    |     from .audio_classification import AudioClassificationPipeline
immich_machine_learning    |   File "/opt/venv/lib/python3.10/site-packages/transformers/pipelines/audio_classification.py", line 22, in <module>
immich_machine_learning    |     from .base import PIPELINE_INIT_ARGS, Pipeline
immich_machine_learning    |   File "/opt/venv/lib/python3.10/site-packages/transformers/pipelines/base.py", line 34, in <module>
immich_machine_learning    |     from ..modelcard import ModelCard
immich_machine_learning    |   File "/opt/venv/lib/python3.10/site-packages/transformers/modelcard.py", line 47, in <module>
immich_machine_learning    |     from .training_args import ParallelMode
immich_machine_learning    |   File "/opt/venv/lib/python3.10/site-packages/transformers/training_args.py", line 28, in <module>
immich_machine_learning    |     from .debug_utils import DebugOption
immich_machine_learning    |   File "/opt/venv/lib/python3.10/site-packages/transformers/debug_utils.py", line 21, in <module>
immich_machine_learning    |     import torch
immich_machine_learning    |   File "/opt/venv/lib/python3.10/site-packages/torch/__init__.py", line 217, in <module>
immich_machine_learning    |     _load_global_deps()
immich_machine_learning    |   File "/opt/venv/lib/python3.10/site-packages/torch/__init__.py", line 178, in _load_global_deps
immich_machine_learning    |     _preload_cuda_deps()
immich_machine_learning    |   File "/opt/venv/lib/python3.10/site-packages/torch/__init__.py", line 158, in _preload_cuda_deps
immich_machine_learning    |     ctypes.CDLL(cublas_path)
immich_machine_learning    | UnboundLocalError: local variable 'cublas_path' referenced before assignment
immich_machine_learning    |
immich_machine_learning    | The above exception was the direct cause of the following exception:
immich_machine_learning    |
immich_machine_learning    | Traceback (most recent call last):
immich_machine_learning    |   File "/usr/src/app/src/main.py", line 3, in <module>
immich_machine_learning    |     from transformers import pipeline
immich_machine_learning    |   File "<frozen importlib._bootstrap>", line 1075, in _handle_fromlist
immich_machine_learning    |   File "/opt/venv/lib/python3.10/site-packages/transformers/utils/import_utils.py", line 1100, in __getattr__
immich_machine_learning    |     module = self._get_module(self._class_to_module[name])
immich_machine_learning    |   File "/opt/venv/lib/python3.10/site-packages/transformers/utils/import_utils.py", line 1112, in _get_module
immich_machine_learning    |     raise RuntimeError(
immich_machine_learning    | RuntimeError: Failed to import transformers.pipelines because of the following error (look up to see its traceback):
immich_machine_learning    | local variable 'cublas_path' referenced before assignment
immich_machine_learning exited with code 1
immich_machine_learning    | Traceback (most recent call last):
immich_machine_learning    |   File "/opt/venv/lib/python3.10/site-packages/torch/__init__.py", line 172, in _load_global_deps
immich_machine_learning    |     ctypes.CDLL(lib_path, mode=ctypes.RTLD_GLOBAL)
immich_machine_learning    |   File "/usr/local/lib/python3.10/ctypes/__init__.py", line 374, in __init__
immich_machine_learning    |     self._handle = _dlopen(self._name, mode)
immich_machine_learning    | OSError: libcublas.so.11: cannot open shared object file: No such file or directory
immich_machine_learning    |
immich_machine_learning    | During handling of the above exception, another exception occurred:
immich_machine_learning    |
immich_machine_learning    | Traceback (most recent call last):
immich_machine_learning    |   File "/opt/venv/lib/python3.10/site-packages/transformers/utils/import_utils.py", line 1110, in _get_module
immich_machine_learning    |     return importlib.import_module("." + module_name, self.__name__)
immich_machine_learning    |   File "/usr/local/lib/python3.10/importlib/__init__.py", line 126, in import_module
immich_machine_learning    |     return _bootstrap._gcd_import(name[level:], package, level)
immich_machine_learning    |   File "<frozen importlib._bootstrap>", line 1050, in _gcd_import
immich_machine_learning    |   File "<frozen importlib._bootstrap>", line 1027, in _find_and_load
immich_machine_learning    |   File "<frozen importlib._bootstrap>", line 1006, in _find_and_load_unlocked
immich_machine_learning    |   File "<frozen importlib._bootstrap>", line 688, in _load_unlocked
immich_machine_learning    |   File "<frozen importlib._bootstrap_external>", line 883, in exec_module
immich_machine_learning    |   File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
immich_machine_learning    |   File "/opt/venv/lib/python3.10/site-packages/transformers/pipelines/__init__.py", line 49, in <module>
immich_machine_learning    |     from .audio_classification import AudioClassificationPipeline
immich_machine_learning    |   File "/opt/venv/lib/python3.10/site-packages/transformers/pipelines/audio_classification.py", line 22, in <module>
immich_machine_learning    |     from .base import PIPELINE_INIT_ARGS, Pipeline
immich_machine_learning    |   File "/opt/venv/lib/python3.10/site-packages/transformers/pipelines/base.py", line 34, in <module>
immich_machine_learning    |     from ..modelcard import ModelCard
immich_machine_learning    |   File "/opt/venv/lib/python3.10/site-packages/transformers/modelcard.py", line 47, in <module>
immich_machine_learning    |     from .training_args import ParallelMode
immich_machine_learning    |   File "/opt/venv/lib/python3.10/site-packages/transformers/training_args.py", line 28, in <module>
immich_machine_learning    |     from .debug_utils import DebugOption
immich_machine_learning    |   File "/opt/venv/lib/python3.10/site-packages/transformers/debug_utils.py", line 21, in <module>
immich_machine_learning    |     import torch
immich_machine_learning    |   File "/opt/venv/lib/python3.10/site-packages/torch/__init__.py", line 217, in <module>
immich_machine_learning    |     _load_global_deps()
immich_machine_learning    |   File "/opt/venv/lib/python3.10/site-packages/torch/__init__.py", line 178, in _load_global_deps
immich_machine_learning    |     _preload_cuda_deps()
immich_machine_learning    |   File "/opt/venv/lib/python3.10/site-packages/torch/__init__.py", line 158, in _preload_cuda_deps
immich_machine_learning    |     ctypes.CDLL(cublas_path)
immich_machine_learning    | UnboundLocalError: local variable 'cublas_path' referenced before assignment
immich_machine_learning    |
immich_machine_learning    | The above exception was the direct cause of the following exception:
immich_machine_learning    |
immich_machine_learning    | Traceback (most recent call last):
immich_machine_learning    |   File "/usr/src/app/src/main.py", line 3, in <module>
immich_machine_learning    |     from transformers import pipeline
immich_machine_learning    |   File "<frozen importlib._bootstrap>", line 1075, in _handle_fromlist
immich_machine_learning    |   File "/opt/venv/lib/python3.10/site-packages/transformers/utils/import_utils.py", line 1100, in __getattr__
immich_machine_learning    |     module = self._get_module(self._class_to_module[name])
immich_machine_learning    |   File "/opt/venv/lib/python3.10/site-packages/transformers/utils/import_utils.py", line 1112, in _get_module
immich_machine_learning    |     raise RuntimeError(
immich_machine_learning    | RuntimeError: Failed to import transformers.pipelines because of the following error (look up to see its traceback):
immich_machine_learning    | local variable 'cublas_path' referenced before assignment
```

**Comment by samip5** [comment]
The issue sounds to come from the fact we don't use the stable PyTorch so it's bound to break sometimes and possibly not related to the changes itself.

**Pull request merged**